### PR TITLE
netbox: 3.5.6 -> 3.5.7 + migration fixes + upgrade NixOS test

### PR DESCRIPTION
**Pull request by minijackson**
## Description of changes

Changelog: https://github.com/netbox-community/netbox/releases/tag/v3.5.7

Some other fixes:

Move migrations into the `preStart` of `netbox.service`, and run the `reindex` command, to allow searches. These migrations are now only run on upgrade/downgrade, or on first run since application of this PR.

Add a `netbox-upgrade` test to test that the upgrade from NetBox 3.3 -> netbox 3.5 works fine. This test can be reused later for other major upgrades.

## Things done

All NetBox-related tests pass (`netbox`, `netbox_3_3`, `netbox-upgrade`)

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>netbox</li>
  </ul>
</details>

---

cc @RaitoBezarius & @n0emis

Let me know if either of you want to be added as maintainer of the `netbox-upgrade` test.

### CODE REVIEW

File: nixos/modules/services/web-apps/netbox.nix [hunk 1]
```diff
@@ -264,40 +264,40 @@ in {
...
-        after = [ "network-online.target" "netbox-migration.service" ];
+        after = [ "network-online.target" ];
         wants = [ "network-online.target" ];
 
+        environment.PYTHONPATH = pkg.pythonPath;
+
         preStart = ''
+          # On the first run, or on upgrade / downgrade, run migrations and related.
+          # This mostly correspond to upstream NetBox's 'upgrade.sh' script.
+          versionFile="/var/lib/netbox/version"
```
Comment (n0emis) [inline nixos/modules/services/web-apps/netbox.nix:281]: Please use `cfg.dataDor here`

**Review by n0emis — APPROVED** [review]
This looks good so far. Do you know how if the preStart script stops, when a single command in it fails? otherwise we should add some error handling

**Comment by minijackson** [comment]
@n0emis what did you have in mind for error handling? I'm more inclined to let the `preStart` script as-is, instead of "silently succeeding" if there's an issue.

**Review by n0emis — APPROVED** [review]
i think this pr is good to go as is

**Pull request merged**